include/uncertainty_planning_core/uncertainty_contact_planning.hpp

```diff
@@ -141,6 +140,100 @@ class UncertaintyPolicyExecutionResult
   }
 };
 
+#if UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 2
+struct ExecutionTimeLimit {
+  ExecutionTimeLimit(double _secs)
+    : secs(_secs), clock(std::make_shared<rclcpp::Clock>())
+  {
+  }
+
+  ExecutionTimeLimit(double _secs, rclcpp::Clock::SharedPtr _clock)
+    : secs(_secs), clock(_clock)
+  {
+  }
+
+  double secs;
+  rclcpp::Clock::SharedPtr clock;
+};
+
+std::ostream& operator<<(
+    std::ostream& strm, const ExecutionTimeLimit& time_limit)
+{
+  return strm << time_limit.secs;
+}
+#elif UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 1
+using ExecutionTimeLimit = double;
+#endif
+
+class ExecutionTimer {
+public:
+#if UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 2
+  inline ExecutionTimer(ExecutionTimeLimit time_limit)
+    : clock_(time_limit.clock), time_limit_(time_limit.secs)
+  {
+  }
+#elif UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 1
+  inline ExecutionTimer(ExecutionTimeLimit time_limit)
+    : time_limit_(time_limit)
+  {
+  }
+#endif
+
+  inline void start()
+  {
+    if (start_time_ >= 0.0) {
+      throw std::logic_error("Execution timer already started");
+    }
+    start_time_ = current_time();
+  }
+
+  inline void stop()
+  {
+    if (stop_time_ >= 0.0) {
+      throw std::logic_error("Execution timer already stopped");
+    }
+    stop_time_ = current_time();
+  }
+
+  inline bool expired() {
+    if (time_limit_ > 0.0)
+    {
+      const double elapsed_time = current_time() - start_time_;
+      return elapsed_time >= time_limit_;
+    }
+    return false;
+  }
+
+  inline double start_time() {
+    return start_time_;
+  }
+
+  inline double stop_time() {
+    return stop_time_;
+  }
+
+  inline double duration() {
+    return stop_time_ - start_time_;
+  }
+
+private:
+#if UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 2
+  double current_time() {
+    return clock_->now().seconds();
+  }
+
+  rclcpp::Clock::SharedPtr clock_;
+#elif UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 1
+  double current_time() {
+    return ros::Time::now().toSec();
+  }
+#endif
+
+  double start_time_{-1.0};
```

Comment:
Since both start and stop time are effectively optionals, let's use `common_robotics_utilities::OwningMaybe<double>` here (if C++17 was used everywhere we'd use `std::optional<double>`).

Reply:
Oh, I didn't know about `common_robotics_utilities::OwningMaybe`. Thanks for the pointer!